drv/opentitan-aontimer/src/lib.rs

```diff
@@ -0,0 +1,154 @@
+// reaches a configurable threshold, it will emit an interrupt signal, called the
+// "bark". If the counter reaches a second configurable threshold, it will reset
+// the system, called the "bite".
+// Enabling the AON Timer before setting a bite threshold causes the system to
+// immediately reset.
+pub struct AonTimer {
+    base_addr: *mut [u32],
+    clock_freq_hz: u32,
+    pub bark_cb: Option<fn()>,
+}
+
+impl AonTimer {
+    // Initialize a new watchdog driver device with the specified base address,
+    // clock frequency (in Hz), and the bark/bite thresholds (both in ms).
+    // If the bark threshold > bite threshold, the bite threshold is set to
+    // the bark threshold.
+    pub fn new(
+        base_addr: u32,
+        clock_freq_hz: u32,
+        bark_thold_ms: u64,
+        bite_thold_ms: u64,
+        bark_cb: Option<fn()>,
+    ) -> Result<Self, ()> {
+        let inst = AonTimer {
+            base_addr: slice_from_raw_parts_mut(base_addr as *mut u32, 12),
+            clock_freq_hz,
+            bark_cb,
+        };
+        unsafe {
+            // This is just a counter, reset it to 0.
+            (*inst.base_addr)[WDOG_COUNT_IDX] = 0;
+            // write 1 to clear any pending interrupts
+            (*inst.base_addr)[INTR_STATE_IDX] =
+                INTR_STATE_BITS_WKUP | INTR_STATE_BITS_WDOG;
+            (*inst.base_addr)[WDOG_BARK_THOLD_IDX] =
+                inst.ms_to_reg_count(bark_thold_ms)?;
+            if bark_thold_ms > bite_thold_ms {
+                return Err(());
+            }
+            (*inst.base_addr)[WDOG_BITE_THOLD_IDX] =
+                inst.ms_to_reg_count(bite_thold_ms)?;
+        }
+        Ok(inst)
+    }
+
+    //Disable the watchdog timer. If the configuration is locked, returns false.
+    pub fn disable(&self) -> Result<(), ()> {
+        if !self.check_lock() {
+            return Err(());
+        }
+        unsafe {
+            // disable everything
+            (*self.base_addr)[WDOG_CTRL_IDX] = 0;
+        }
+        Ok(())
+    }
+
+    //Enables the watchdog timer. If the configuration is locked, returns false.
+    pub fn enable(&self) -> Result<(), ()> {
+        if !self.check_lock() {
+            return Err(());
+        }
+        unsafe {
+            (*self.base_addr)[WDOG_CTRL_IDX] |= WDOG_CTRL_BITS_EN;
+        }
+        Ok(())
+    }
+
+    //Enables the watchdog timer, then locks the configuration. If the
+    // configuration is already locked, returns false.
+    pub fn enable_and_lock(&self) -> Result<(), ()> {
+        if !self.check_lock() {
+            return Err(());
+        }
+        unsafe {
+            (*self.base_addr)[WDOG_CTRL_IDX] |= WDOG_CTRL_BITS_EN;
+            (*self.base_addr)[WDOG_REGWEN_IDX] = 0;
+        }
+        Ok(())
+    }
+
+    //Sets the bark threshold in ms. If the configuration is locked, returns false.
+    pub fn set_bark_thold(&self, bark_thold_ms: u64) -> Result<(), ()> {
```

Comment:
Just noting that there are no checks in `set_bark_threshold` and `set_bite_threshold` to ensure that bark and bite thresholds are not configured incorrectly w.r.t. each other. Do we need to expose these APIs at all?

Reply:
That is a fair point, although it is a valid configuration as far as the AON timer is concerned.  I think the check within the task that uses the driver is good enough.  
I will remove the `pub` for now.